File: contrib/apache-orc.bak.1766388422/java/core/src/java/org/apache/orc/util/BloomFilter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.orc.util;

import java.nio.charset.Charset;
import java.util.Arrays;

/**
 * BloomFilter is a probabilistic data structure for set membership check. BloomFilters are
 * highly space efficient when compared to using a HashSet. Because of the probabilistic nature of
 * bloom filter false positive (element not present in bloom filter but test() says true) are
 * possible but false negatives are not possible (if element is present then test() will never
 * say false). The false positive probability is configurable (default: 5%) depending on which
 * storage requirement may increase or decrease. Lower the false positive probability greater
 * is the space requirement.
 * Bloom filters are sensitive to number of elements that will be inserted in the bloom filter.
 * During the creation of bloom filter expected number of entries must be specified. If the number
 * of insertions exceed the specified initial number of entries then false positive probability will
 * increase accordingly.
 * <p>
 * Internally, this implementation of bloom filter uses Murmur3 fast non-cryptographic hash
 * algorithm. Although Murmur2 is slightly faster than Murmur3 in Java, it suffers from hash
 * collisions for specific sequence of repeating bytes. Check the following link for more info
 * https://code.google.com/p/smhasher/wiki/MurmurHash2Flaw
 * <p>
 * Note that this class is here for backwards compatibility, because it uses
 * the JVM default character set for strings. All new users should
 * BloomFilterUtf8, which always uses UTF8 for the encoding.
 */
public class BloomFilter {
  public static final double DEFAULT_FPP = 0.05;
  private final BitSet bitSet;
  private final int numBits;
  private final int numHashFunctions;

  static void checkArgument(boolean expression, String message) {
    if (!expression) {
      throw new IllegalArgumentException(message);
    }
  }

  public BloomFilter(long expectedEntries) {
    this(expectedEntries, DEFAULT_FPP);
  }

  public BloomFilter(long expectedEntries, double fpp) {
    expectedEntries = Math.max(expectedEntries, 1);
    checkArgument(fpp > 0.0 && fpp < 1.0, "False positive probability should be > 0.0 & < 1.0");
    int nb = optimalNumOfBits(expectedEntries, fpp);
    // make 'm' multiple of 64
    this.numBits = nb + (Long.SIZE - (nb % Long.SIZE));
    this.numHashFunctions = optimalNumOfHashFunctions(expectedEntries, numBits);
    this.bitSet = new BitSet(numBits);
  }

  /**
   * A constructor to support rebuilding the BloomFilter from a serialized representation.
   * @param bits the serialized bits
   * @param numFuncs the number of functions used
   */
  public BloomFilter(long[] bits, int numFuncs) {
    super();
    bitSet = new BitSet(bits);
    this.numBits = (int) bitSet.bitSize();
    numHashFunctions = numFuncs;
  }

  static int optimalNumOfHashFunctions(long n, long m) {
    return Math.max(1, (int) Math.round((double) m / n * Math.log(2)));
  }

  static int optimalNumOfBits(long n, double p) {
    return (int) (-n * Math.log(p) / (Math.log(2) * Math.log(2)));
  }

  @Override
  public boolean equals(Object other) {
    return other != null &&
        other.getClass() == getClass() &&
        numBits == ((BloomFilter) other).numBits &&
        numHashFunctions == ((BloomFilter) other).numHashFunctions &&
        bitSet.equals(((BloomFilter) other).bitSet);
  }

  @Override
  public int hashCode() {
    return bitSet.hashCode() + numHashFunctions * 5;
  }

  public void add(byte[] val) {
    addBytes(val, 0, val == null ? 0 : val.length);
  }

  public void addBytes(byte[] val, int offset, int length) {
    // We use the trick mentioned in "Less Hashing, Same Performance: Building a Better Bloom Filter"
    // by Kirsch et.al. From abstract 'only two hash functions are necessary to effectively
    // implement a Bloom filter without any loss in the asymptotic false positive probability'

    // Lets split up 64-bit hashcode into two 32-bit hash codes and employ the technique mentioned
    // in the above paper
    long hash64 = val == null ? Murmur3.NULL_HASHCODE :
        Murmur3.hash64(val, offset, length);
    addHash(hash64);
  }

  private void addHash(long hash64) {
    int hash1 = (int) hash64;
    int hash2 = (int) (hash64 >>> 32);

    for (int i = 1; i <= numHashFunctions; i++) {
      int combinedHash = hash1 + (i * hash2);
      // hashcode should be positive, flip all the bits if it's negative
      if (combinedHash < 0) {
        combinedHash = ~combinedHash;
      }
      int pos = combinedHash % numBits;
      bitSet.set(pos);
    }
  }

  public void addString(String val) {
    if (val == null) {
      add(null);
    } else {
      add(val.getBytes(Charset.defaultCharset()));
    }
  }

  public void addLong(long val) {
    addHash(getLongHash(val));
  }

  public void addDouble(double val) {
    addLong(Double.doubleToLongBits(val));
  }

  public boolean test(byte[] val) {
    return testBytes(val, 0, val == null ? 0 : val.length);
  }

  public boolean testBytes(byte[] val, int offset, int length) {
    long hash64 = val == null ? Murmur3.NULL_HASHCODE :
        Murmur3.hash64(val, offset, length);
    return testHash(hash64);
  }

  private boolean testHash(long hash64) {
    int hash1 = (int) hash64;
    int hash2 = (int) (hash64 >>> 32);

    for (int i = 1; i <= numHashFunctions; i++) {
      int combinedHash = hash1 + (i * hash2);
      // hashcode should be positive, flip all the bits if it's negative
      if (combinedHash < 0) {
        combinedHash = ~combinedHash;
      }
      int pos = combinedHash % numBits;
      if (!bitSet.get(pos)) {
        return false;
      }
    }
    return true;
  }

  public boolean testString(String val) {
    if (val == null) {
      return test(null);
    } else {
      return test(val.getBytes(Charset.defaultCharset()));
    }
  }

  public boolean testLong(long val) {
    return testHash(getLongHash(val));
  }

  // Thomas Wang's integer hash function
  // http://web.archive.org/web/20071223173210/http://www.concentric.net/~Ttwang/tech/inthash.htm
  static long getLongHash(long key) {
    key = (~key) + (key << 21); // key = (key << 21) - key - 1;
    key = key ^ (key >> 24);
    key = (key + (key << 3)) + (key << 8); // key * 265
    key = key ^ (key >> 14);
    key = (key + (key << 2)) + (key << 4); // key * 21
    key = key ^ (key >> 28);
    key = key + (key << 31);
    return key;
  }

  public boolean testDouble(double val) {
    return testLong(Double.doubleToLongBits(val));
  }

  public long sizeInBytes() {
    return getBitSize() / 8;
  }

  public int getBitSize() {
    return bitSet.getData().length * Long.SIZE;
  }

  public int getNumHashFunctions() {
    return numHashFunctions;
  }

  public long[] getBitSet() {
    return bitSet.getData();
  }

  @Override
  public String toString() {
    return "m: " + numBits + " k: " + numHashFunctions;
  }

  /**
   * Merge the specified bloom filter with current bloom filter.
   *
   * @param that - bloom filter to merge
   */
  public void merge(BloomFilter that) {
    if (this != that && numBits == that.numBits && numHashFunctions == that.numHashFunctions) {
      this.bitSet.putAll(that.bitSet);
    } else {
      throw new IllegalArgumentException("BloomFilters are not compatible for merging." +
          " this - " + this + " that - " + that);
    }
  }

  public void reset() {
    this.bitSet.clear();
  }

  /**
   * Helper method that only used for tests. Check if the given position in the bitSet is
   * true. Use default visibility.
   */
  boolean testBitSetPos(int pos) {
    return this.bitSet.get(pos);
  }

  /**
   * Bare metal bit set implementation. For performance reasons, this implementation does not check
   * for index bounds nor expand the bit set size if the specified index is greater than the size.
   */
  public static class BitSet {
    private final long[] data;

    public BitSet(long bits) {
      this(new long[(int) Math.ceil((double) bits / (double) Long.SIZE)]);
    }

    /**
     * Deserialize long array as bit set.
     *
     * @param data - bit array
     */
    public BitSet(long[] data) {
      assert data.length > 0 : "data length is zero!";
      this.data = data;
    }

    /**
     * Sets the bit at specified index.
     *
     * @param index - position
     */
    public void set(int index) {
      data[index >>> 6] |= (1L << index);
    }

    /**
     * Returns true if the bit is set in the specified index.
     *
     * @param index - position
     * @return - value at the bit position
     */
    public boolean get(int index) {
      return (data[index >>> 6] & (1L << index)) != 0;
    }

    /**
     * Number of bits
     */
    public long bitSize() {
      return (long) data.length * Long.SIZE;
    }

    public long[] getData() {
      return data;
    }

    /**
     * Combines the two BitArrays using bitwise OR.
     */
    public void putAll(BitSet array) {
      assert data.length == array.data.length :
          "BitArrays must be of equal length (" + data.length + "!= " + array.data.length + ")";
      for (int i = 0; i < data.length; i++) {
        data[i] |= array.data[i];
      }
    }

    /**
     * Clear the bit set.
     */
    public void clear() {
      Arrays.fill(data, 0);
    }

    @Override
    public boolean equals(Object other) {
      return other != null &&
          other.getClass() == getClass() &&
          Arrays.equals(data, ((BitSet) other).data);
    }

    @Override
    public int hashCode() {
      return Arrays.hashCode(data);
    }
  }
}

File: contrib/apache-orc.bak.1766388422/c++/src/BloomFilter.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "BloomFilter.hh"
#include "Murmur3.hh"

namespace orc {

  constexpr uint64_t BITS_OF_LONG = 64;
  constexpr uint8_t SHIFT_6_BITS = 6;
  constexpr uint8_t SHIFT_3_BITS = 3;

  static bool isLittleEndian() {
    static union {
      uint32_t i;
      char c[4];
    } num = {0x01020304};
    return num.c[0] == 4;
  }

  /**
   * Implementation of BitSet
   */
  BitSet::BitSet(uint64_t numBits) {
    mData.resize(static_cast<size_t>(ceil(static_cast<double>(numBits) / BITS_OF_LONG)), 0);
  }

  BitSet::BitSet(const uint64_t* bits, uint64_t numBits) {
    // caller should make sure numBits is multiple of 64
    mData.resize(numBits >> SHIFT_6_BITS, 0);
    memcpy(mData.data(), bits, numBits >> SHIFT_3_BITS);
  }

  void BitSet::set(uint64_t index) {
    mData[index >> SHIFT_6_BITS] |= (1ULL << (index % BITS_OF_LONG));
  }

  bool BitSet::get(uint64_t index) {
    return (mData[index >> SHIFT_6_BITS] & (1ULL << (index % BITS_OF_LONG))) != 0;
  }

  uint64_t BitSet::bitSize() {
    return mData.size() << SHIFT_6_BITS;
  }

  void BitSet::merge(const BitSet& other) {
    if (mData.size() != other.mData.size()) {
      std::stringstream ss;
      ss << "BitSet must be of equal length (" << mData.size() << " != " << other.mData.size()
         << ")";
      throw std::logic_error(ss.str());
    }

    for (size_t i = 0; i != mData.size(); i++) {
      mData[i] |= other.mData[i];
    }
  }

  void BitSet::clear() {
    memset(mData.data(), 0, sizeof(uint64_t) * mData.size());
  }

  const uint64_t* BitSet::getData() const {
    return mData.data();
  }

  bool BitSet::operator==(const BitSet& other) const {
    return mData == other.mData;
  }

  /**
   * Helper functions
   */
  void checkArgument(bool expression, const std::string& message) {
    if (!expression) {
      throw std::logic_error(message);
    }
  }

  int32_t optimalNumOfHashFunctions(uint64_t expectedEntries, uint64_t numBits) {
    double n = static_cast<double>(expectedEntries);
    return std::max<int32_t>(
        1, static_cast<int32_t>(std::round(static_cast<double>(numBits) / n * std::log(2.0))));
  }

  int32_t optimalNumOfBits(uint64_t expectedEntries, double fpp) {
    double n = static_cast<double>(expectedEntries);
    return static_cast<int32_t>(-n * std::log(fpp) / (std::log(2.0) * std::log(2.0)));
  }

  // We use the trick mentioned in "Less Hashing, Same Performance:
  // Building a Better Bloom Filter" by Kirsch et.al. From abstract
  // 'only two hash functions are necessary to effectively implement
  // a Bloom filter without any loss in the asymptotic false positive
  // probability'
  // Lets split up 64-bit hashcode into two 32-bit hash codes and employ
  // the technique mentioned in the above paper
  inline uint64_t getBytesHash(const char* data, int64_t length) {
    if (data == nullptr) {
      return Murmur3::NULL_HASHCODE;
    }

    return Murmur3::hash64(reinterpret_cast<const uint8_t*>(data), static_cast<uint32_t>(length));
  }

  /**
   * Implementation of BloomFilter
   */
  BloomFilterImpl::BloomFilterImpl(uint64_t expectedEntries, double fpp) {
    checkArgument(expectedEntries > 0, "expectedEntries should be > 0");
    checkArgument(fpp > 0.0 && fpp < 1.0, "False positive probability should be > 0.0 & < 1.0");

    uint64_t nb = static_cast<uint64_t>(optimalNumOfBits(expectedEntries, fpp));
    // make 'mNumBits' multiple of 64
    mNumBits = nb + (BITS_OF_LONG - (nb % BITS_OF_LONG));
    mNumHashFunctions = optimalNumOfHashFunctions(expectedEntries, mNumBits);
    mBitSet.reset(new BitSet(mNumBits));
  }

  void BloomFilterImpl::addBytes(const char* data, int64_t length) {
    uint64_t hash64 = getBytesHash(data, length);
    addHash(static_cast<int64_t>(hash64));
  }

  void BloomFilterImpl::addLong(int64_t data) {
    addHash(getLongHash(data));
  }

  bool BloomFilterImpl::testBytes(const char* data, int64_t length) const {
    uint64_t hash64 = getBytesHash(data, length);
    return testHash(static_cast<int64_t>(hash64));
  }

  bool BloomFilterImpl::testLong(int64_t data) const {
    return testHash(getLongHash(data));
  }

  uint64_t BloomFilterImpl::sizeInBytes() const {
    return getBitSize() >> SHIFT_3_BITS;
  }

  uint64_t BloomFilterImpl::getBitSize() const {
    return mBitSet->bitSize();
  }

  int32_t BloomFilterImpl::getNumHashFunctions() const {
    return mNumHashFunctions;
  }

  DIAGNOSTIC_PUSH

#if defined(__clang__)
  DIAGNOSTIC_IGNORE("-Wundefined-reinterpret-cast")
#endif

#if defined(__GNUC__)
  DIAGNOSTIC_IGNORE("-Wstrict-aliasing")
#endif

  // caller should make sure input proto::BloomFilter is valid since
  // no check will be performed in the following constructor
  BloomFilterImpl::BloomFilterImpl(const proto::BloomFilter& bloomFilter) {
    mNumHashFunctions = static_cast<int32_t>(bloomFilter.numhashfunctions());

    const std::string& bitsetStr = bloomFilter.utf8bitset();
    mNumBits = bitsetStr.size() << SHIFT_3_BITS;
    checkArgument(mNumBits % BITS_OF_LONG == 0, "numBits should be multiple of 64!");

    const uint64_t* bitset = reinterpret_cast<const uint64_t*>(bitsetStr.data());
    if (isLittleEndian()) {
      mBitSet.reset(new BitSet(bitset, mNumBits));
    } else {
      std::vector<uint64_t> longs(mNumBits >> SHIFT_6_BITS);
      for (size_t i = 0; i != longs.size(); ++i) {
        // convert little-endian to big-endian
        const uint64_t src = bitset[i];
        uint64_t& dst = longs[i];
        for (size_t bit = 0; bit != 64; bit += 8) {
          dst |= (((src & (0xFFu << bit)) >> bit) << (56 - bit));
        }
      }

      mBitSet.reset(new BitSet(longs.data(), mNumBits));
    }
  }

  void BloomFilterImpl::addDouble(double data) {
    addLong(reinterpret_cast<int64_t&>(data));
  }

  bool BloomFilterImpl::testDouble(double data) const {
    return testLong(reinterpret_cast<int64_t&>(data));
  }

  DIAGNOSTIC_POP

  void BloomFilterImpl::addHash(int64_t hash64) {
    int32_t hash1 = static_cast<int32_t>(hash64 & 0xffffffff);
    // In Java codes, we use "hash64 >>> 32" which is an unsigned shift op.
    // So we cast hash64 to uint64_t here for an unsigned right shift.
    int32_t hash2 = static_cast<int32_t>(static_cast<uint64_t>(hash64) >> 32);

    for (int32_t i = 1; i <= mNumHashFunctions; ++i) {
      int32_t combinedHash = hash1 + i * hash2;
      // hashcode should be positive, flip all the bits if it's negative
      if (combinedHash < 0) {
        combinedHash = ~combinedHash;
      }
      uint64_t pos = static_cast<uint64_t>(combinedHash) % mNumBits;
      mBitSet->set(pos);
    }
  }

  bool BloomFilterImpl::testHash(int64_t hash64) const {
    int32_t hash1 = static_cast<int32_t>(hash64 & 0xffffffff);
    // In Java codes, we use "hash64 >>> 32" which is an unsigned shift op.
    // So we cast hash64 to uint64_t here for an unsigned right shift.
    int32_t hash2 = static_cast<int32_t>(static_cast<uint64_t>(hash64) >> 32);

    for (int32_t i = 1; i <= mNumHashFunctions; ++i) {
      int32_t combinedHash = hash1 + i * hash2;
      // hashcode should be positive, flip all the bits if it's negative
      if (combinedHash < 0) {
        combinedHash = ~combinedHash;
      }
      uint64_t pos = static_cast<uint64_t>(combinedHash) % mNumBits;
      if (!mBitSet->get(pos)) {
        return false;
      }
    }
    return true;
  }

  void BloomFilterImpl::merge(const BloomFilterImpl& other) {
    if (mNumBits != other.mNumBits || mNumHashFunctions != other.mNumHashFunctions) {
      std::stringstream ss;
      ss << "BloomFilters are not compatible for merging: "
         << "this: numBits:" << mNumBits << ",numHashFunctions:" << mNumHashFunctions
         << ", that: numBits:" << other.mNumBits << ",numHashFunctions:" << other.mNumHashFunctions;
      throw std::logic_error(ss.str());
    }

    mBitSet->merge(*other.mBitSet);
  }

  void BloomFilterImpl::reset() {
    mBitSet->clear();
  }

  void BloomFilterImpl::serialize(proto::BloomFilter& bloomFilter) const {
    bloomFilter.set_numhashfunctions(static_cast<uint32_t>(mNumHashFunctions));

    // According to ORC standard, the encoding is a sequence of bytes with
    // a little endian encoding in the utf8bitset field.
    if (isLittleEndian()) {
      // bytes are already organized in little endian; thus no conversion needed
      const char* bitset = reinterpret_cast<const char*>(mBitSet->getData());
      bloomFilter.set_utf8bitset(bitset, sizeInBytes());
    } else {
      std::vector<uint64_t> bitset(sizeInBytes() / sizeof(uint64_t), 0);
      const uint64_t* longs = mBitSet->getData();
      for (size_t i = 0; i != bitset.size(); ++i) {
        uint64_t& dst = bitset[i];
        const uint64_t src = longs[i];
        // convert big-endian to little-endian
        for (size_t bit = 0; bit != 64; bit += 8) {
          dst |= (((src & (0xFFu << bit)) >> bit) << (56 - bit));
        }
      }
      bloomFilter.set_utf8bitset(bitset.data(), sizeInBytes());
    }
  }

  bool BloomFilterImpl::operator==(const BloomFilterImpl& other) const {
    return mNumBits == other.mNumBits && mNumHashFunctions == other.mNumHashFunctions &&
           *mBitSet == *other.mBitSet;
  }

  BloomFilter::~BloomFilter() {
    // PASS
  }

  std::unique_ptr<BloomFilter> BloomFilterUTF8Utils::deserialize(
      const proto::Stream_Kind& streamKind, const proto::ColumnEncoding& encoding,
      const proto::BloomFilter& bloomFilter) {
    // only BLOOM_FILTER_UTF8 is supported
    if (streamKind != proto::Stream_Kind_BLOOM_FILTER_UTF8) {
      return nullptr;
    }

    // make sure we don't use unknown encodings or original timestamp encodings
    if (!encoding.has_bloomencoding() || encoding.bloomencoding() != 1) {
      return nullptr;
    }

    // make sure all required fields exist
    if (!bloomFilter.has_numhashfunctions() || !bloomFilter.has_utf8bitset()) {
      return nullptr;
    }

    return std::make_unique<BloomFilterImpl>(bloomFilter);
  }

}  // namespace orc
